src/website/webview/views.py

```diff
@@ -1,16 +1,363 @@
+                            ),
+                        ),
+                        default=Value(0.0),
+                    ),
+                    rank_description=Case(
+                        When(
+                            Q(metadata__search_vector=search_pkgs),
+                            then=SearchRank(
+                                F("metadata__search_vector"),
+                                search_query,
+                                normalization=norm_value,
+                            ),
+                        ),
+                        default=Value(0.0),
+                    ),
+                )
+                .values("name")
+                .annotate(
+                    pkg_count=Count("name"),
+                    ids=ArrayAgg("id", ordering="id"),
+                    attributes=ArrayAgg("attribute", ordering="id"),
+                    description=Max("metadata__description"),
+                    max_rank_pkg=Max("rank_pkg"),
+                    max_rank_description=Max("rank_description"),
+                )
+                .order_by(
+                    "-max_rank_pkg",
+                    "-max_rank_description",
+                    "name",
+                )
+            )
+
+            # There's no need for a separate efficient query in the case of pkg aggregation
+            # because the related field is a o2m instead of several m2m related fields.
+            pkg_count_function = pkg_qs.count
+
+            return pkg_qs, pkg_count_function
+        else:
+            pkg_qs = (
+                NixDerivation.objects.values("name")
+                .annotate(
+                    pkg_count=Count("name"),
+                    ids=ArrayAgg("id", ordering="id"),
+                    attributes=ArrayAgg("attribute", ordering="id"),
+                    description=Max("metadata__description"),
+                )
+                .order_by("name")
+            )
+            pkg_count_function = NixDerivation.objects.values("name").distinct().count
+
+            return pkg_qs, pkg_count_function
+
+    def _get_cve_page(
+        self, search_cves: str | None, cve_page_number: int
+    ) -> tuple[Page, CustomCountPaginator]:
+        cve_qs, cve_count_function = self._get_cve_qs(search_cves)
+        cve_paginator = CustomCountPaginator(
+            cve_qs, self.paginate_by, custom_count=cve_count_function
+        )
+        cve_page = cve_paginator.get_page(cve_page_number)
+        return cve_page, cve_paginator
+
+    def _get_pkg_page(
+        self, search_pkgs: str | None, pkg_page_number: int
+    ) -> tuple[Page, CustomCountPaginator]:
+        pkg_qs, pkg_count_function = self._get_pkg_qs(search_pkgs)
+        pkg_paginator = CustomCountPaginator(
+            pkg_qs, self.paginate_by, custom_count=pkg_count_function
+        )
+        pkg_page = pkg_paginator.get_page(pkg_page_number)
+        return pkg_page, pkg_paginator
```

Comment:
Not critical: One could pass what's currently `_get_cve_qs` and `_get_pkgs_qs` as a parameter `_get_queryset`, and rename the search string and page number variables to be more generic. But it's okay for now, the duplication is not super confusing.